teps/0092-scheduling-timeout.md

```diff
@@ -17,361 +22,470 @@ authors:
-* `scheduling`
 
-Instead of creating a new nested section, we could create a top level section, prefacing the option name with
-`taskruns-` to try to indicate that this option - unlike the existing timeout configuration - is applied to _every_
-`taskrun`:
+#### Invalid Cases
 
 ```yaml
 spec:
   timeouts:
-    pipeline: "0h15m0s"
-    tasks: "0h10m0s"
-    taskrun-scheduling: "5m" # this is the new option
-    finally: "0h5m0s"
+    scheduling: "0m"
+    total: "20m"
 ```
+Invalid: Since `scheduling` timeout is set to 0, `total` timeout must also be set to 0. This will result in a validation error.
 
-This feels like it's trying to accomplish the same thing as the proposed syntax but is less clear (the information
-is encoded in the name of the string vs. the structure of the option) and does not support adding more taskrun level
-options as well (any new taskrun level options would also need to be prefaced with `taskrun-`).
+```yaml
+spec:
+  timeouts:
+    execution: "0m"
+    total: "20m"
+```
+Invalid: Since `exection` timeout is set to 0, `total` timeout must also be set to 0.
 
-#### The field in TaskRuns that represents the existing timeout
 
-In the proposal above, we replace the existing `timeout` field in `TaskRun` with a dictionary (supporting both at least
-for the moment), and use `execution` as the name of the field that does the same thing as the existing field:
+```yaml
+spec:
+  timeouts:
+    scheduling: "5m"
+    execution: "10"
+    total: "20m"
+```
+Invalid: This is invalid because `scheduling` and `execution` timeout does not add up to `total` timeout.
 
 ```yaml
 spec:
   timeouts:
-    scheduling: "5m" # this is the new option
-    execution: "10m" # this option would replace the existing `timeout` field
+    scheduling: "10m"
+    execution: "75m"
+    total: "0m"
 ```
+Invalid: This is invalid because `scheduling` and `execution` timeout does not add up to `total` timeout.
 
-Here is a list of some possible names for this option, comment if you prefer any of these (or have other ideas):
+### What if the pod gets rescheduled?
 
-* `execution`
-* `run`
-* `completion`
+If the `Pod` gets rescheduled, the `scheduling` timer would not restart, i.e. it would still be considered already scheduled.
 
-## Test Plan
+As a side note, we already do not handle this situation well, see [pipelines#2813](https://github.com/tektoncd/pipeline/issues/2813).
+
+
+### PipelineRun
+
+#### TaskRunTemplate
+
+In [TEP-0119][tep-0119], we added `TaskRunTemplate` to support setting configuration for all `TaskRun` in `PipelineRun`.  This proposal adds a new timeout configuration within a `TaskRunTemplate` that allows users to specify `scheduling`, `execution` and `total` timeout for all `TaskRuns` executed in a given `PipelineRun`.
+
+```yaml=
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: clone-test-build-
+spec:
+  pipelineRef:
+    name: clone-test-build
+  taskRunTemplate:
```

Comment:
I'm curious what would happen if you used this TaskRun template in addition to pipeline task timeouts. This could create some confusing results. e.g. with the below template, using a pipeline task timeout of 5m would be invalid.

Reply:
I think that runtime specifications should take precedence over authoring time specifications

taskrunspecs > taskruntemplate > pipelinetask 

but we'd have to be clear about it in the documentation

wonder what happens if retries are set in both taskrunspecs/taskruntemplate and pipelinetask 🤔 

also open to making this scenario validation error initially